File: javascript/leap/leap.js
'use strict';
function divisibleBy(divisor) {
  return (dividend) => dividend % divisor === 0;
}

function not(fn) {
  return (x) => !fn(x);
}

function and(f, g) {
  return (x) => f(x) && g(x);
}

function or(f, g) {
  return (x) => f(x) || g(x);
}

const leap = or(
  and(divisibleBy(4), not(divisibleBy(100))),
  divisibleBy(400)
);

module.exports = leap;

File: python/leap/leap.py
def divisible_by(divisor):
    return lambda dividend: dividend % divisor == 0


def negation(fn):
    return lambda x: not fn(x)


def both(f, g):
    return lambda x: f(x) and g(x)


def either(f, g):
    return lambda x: f(x) or g(x)


def is_leap_year(year):
    return either(
        both(
            divisible_by(4), negation(divisible_by(100))),
        divisible_by(400))(year)
